src/native.rs

```diff
@@ -1414,6 +1415,164 @@ impl super::Context for Context {
+        );
+    }
+
+    unsafe fn debug_message_callback(
+        &self,
+        callback: Self::DebugProc,
+        user_data: *mut std::ffi::c_void,
+    ) {
+        let gl = &self.raw;
+        gl.DebugMessageCallback(callback, user_data);
+    }
+
+    unsafe fn get_debug_message_log(&self, count: u32) -> Vec<DebugMessageLogEntry> {
+        let ct = count as usize;
+        let mut sources = Vec::with_capacity(ct);
+        let mut types = Vec::with_capacity(ct);
+        let mut ids = Vec::with_capacity(ct);
+        let mut severities = Vec::with_capacity(ct);
+        let mut lengths = Vec::with_capacity(ct);
+        let buf_size = (count * MAX_DEBUG_MESSAGE_LENGTH) as i32;
+        let mut message_log = Vec::with_capacity(buf_size as usize);
+
+        let gl = &self.raw;
+        let received = gl.GetDebugMessageLog(
+            count,
+            buf_size,
+            sources.as_mut_ptr(),
+            types.as_mut_ptr(),
+            ids.as_mut_ptr(),
+            severities.as_mut_ptr(),
+            lengths.as_mut_ptr(),
+            message_log.as_mut_ptr(),
+        ) as usize;
+        println!("{} messages received", received);
+
+        sources.set_len(received);
+        types.set_len(received);
+        ids.set_len(received);
+        severities.set_len(received);
+        lengths.set_len(received);
+        message_log.set_len(buf_size as usize);
+
+        let mut entries = Vec::new();
+        let mut offset = 0;
+        for i in 0..received {
+            let message = std::ffi::CStr::from_ptr(message_log[offset..].as_ptr()).to_string_lossy();
+            offset += lengths[i] as usize;
+            entries.push(DebugMessageLogEntry {
+                source: sources[i],
+                msg_type: types[i],
+                id: ids[i],
+                severity: severities[i],
+                message: message.to_string(),
+            });
+        }
+
+        entries
+    }
+
+    unsafe fn get_pointerv(&self, pname: u32, params: *const *mut std::ffi::c_void) {
+        let gl = &self.raw;
+        gl.GetPointerv(pname, params);
+    }
+
+    unsafe fn push_debug_group<S>(&self, source: u32, id: u32, message: S) where S: AsRef<str> {
+        let gl = &self.raw;
+        let msg = CString::new(message.as_ref()).unwrap();
+        gl.PushDebugGroup(source, id, msg.to_bytes().len() as i32, msg.as_ptr());
+    }
+
+    unsafe fn pop_debug_group(&self) {
+        let gl = &self.raw;
+        gl.PopDebugGroup();
+    }
+
+    unsafe fn object_label<S>(&self, identifier: u32, name: u32, label: S) where S: AsRef<str> {
+        let gl = &self.raw;
+        let lbl = CString::new(label.as_ref()).unwrap();
+
+        gl.ObjectLabel(identifier, name, lbl.to_bytes().len() as i32, lbl.as_ptr());
+    }
+
+    unsafe fn get_object_label(&self, identifier: u32, name: u32) -> String {
+        let gl = &self.raw;
+        let mut max_len = 0;
+        gl.GetIntegerv(MAX_LABEL_LENGTH, &mut max_len);
```

Comment:
somewhat loosely related: I wonder if constants like this are better to just check at context creation and keep in `self` forever instead of re-querying on every call.